File: Scala/src/com/yourtion/Pattern18/LazySequenceExample.scala
package com.yourtion.Pattern18

import scala.util.Random

/**
  * Created by Yourtion on 06/04/2017.
  */
object LazySequenceExample {

  val integers = Stream.from(0)

  val generate = new Random()
  val randoms = Stream.continually(generate.nextInt)

  def pagedSequence(pageNum: Int): Stream[String] =
    getPage(pageNum) match {
      case Some(page: String) => page #:: pagedSequence(pageNum + 1)
      case None => Stream.Empty
    }

  def getPage(page: Int) =
    page match {
      case 1 => Some("Page1")
      case 2 => Some("Page2")
      case 3 => Some("Page3")
      case _ => None
    }

  def run(): Unit = {
    println("Some Ints :")
    val someints = integers take 5
    someints foreach println
    println()

    println("aFewRabdom :")
    val aFewRabdom = randoms take 5
    aFewRabdom foreach println
    println()

    println("aFewMoreRabdom :")
    val aFewMoreRabdom = randoms take 6
    aFewMoreRabdom foreach println
    println()

    println("print Hello :")
    val printHellos = Stream.continually(println("hello"))
    println(printHellos take 5 )
    println()
    println(printHellos take 5 force)
    println()

    println("aStream head :")
    val aStream = "foo" #:: "bar" #:: Stream[String]()
    println("Head:" + aStream.head)
    println("Tail: " + aStream.tail)
    println()

    println("pages :")
    println(pagedSequence(1) take 2 force)
    println(pagedSequence(1) force)
    println()

    val holdsHead = {
      def pagedSequence(pageNum: Int): Stream[String] =
        getPage(pageNum) match {
          case Some(page: String) => {
            println("Realizing " + page)
            page #:: pagedSequence(pageNum + 1)
          }
          case None => Stream.Empty
        }
      pagedSequence(1)
    }

    println("holdsHead :")
    println(holdsHead force)
    println()

    def doesntHoldHead = {
      def pagedSequence(pageNum: Int): Stream[String] =
        getPage(pageNum) match {
          case Some(page: String) => {
            println("Realizing " + page)
            page #:: pagedSequence(pageNum + 1)
          }
          case None => Stream.Empty
        }
      pagedSequence(1)
    }
    println("doesntHoldHead :")
    println(doesntHoldHead force)
    println()
  }
}

File: Clojure/src/com/yourtion/Pattern18/lazy_sequence_example.clj
(ns com.yourtion.Pattern18.lazy-sequence-example)

(def integers (range Integer/MAX_VALUE))

(def randoms (repeatedly (fn [] (rand-int Integer/MAX_VALUE))))

(defn get-page [page-num]
  (cond
    (= page-num 1) "Page1"
    (= page-num 2) "Page2"
    (= page-num 3) "Page3"
    :default nil))

(defn paged-sequence [page-num]
  (let [page (get-page page-num)]
    (when page
      (cons page (lazy-seq (paged-sequence (inc page-num)))))))

(defn paged-sequence-print [page-num]
  (let [page (get-page page-num)]
    (when page
      (println (str "Realizing " page))
      (cons page (lazy-seq (paged-sequence (inc page-num)))))))

(defn print-num [num] (print (str num " ")))

(def print-hellos (repeatedly (fn [] (println "hello, world"))))

(defn run []
  (println "Some Ints :")
  (println (take 5 integers))
  (println "")

  (println "aFewRabdom :")
  (println (take 5 randoms))
  (println "")

  (println "aFewMoreRabdom :")
  (println (take 6 randoms))
  (println "")

  (println "print Hello :")
  (println (take 5 print-hellos))
  (println "")

  (println "pages :")
  (println (paged-sequence 1))
  (println (take 2 (paged-sequence 1)))

  (println ""))
